src/geometry/euclidean.lean

```diff
@@ -0,0 +1,98 @@
+/-
+Copyright (c) 2020 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Joseph Myers.
+-/
+import algebra.big_operators
+import analysis.normed_space.real_inner_product
+import analysis.normed_space.add_torsor
+
+noncomputable theory
+open_locale big_operators
+
+/-!
+# Euclidean spaces
+
+This file defines Euclidean affine spaces.
+
+## Implementation notes
+
+Rather than requiring Euclidean affine spaces to be finite-dimensional
+(as in the definition on Wikipedia), this is specified only for those
+theorems that need it.
+
+## References
+
+* https://en.wikipedia.org/wiki/Euclidean_space
+
+-/
+
+section prio
+set_option default_priority 100 -- see Note [default priority]
+/-- A `euclidean_affine_space V P` is an affine space with points `P`
+over an `inner_product_space V`. -/
+class euclidean_affine_space (V : Type*) (P : Type*) [inner_product_space V] [metric_space P]
+  extends normed_add_torsor V P
+end prio
+
+/-- The standard Euclidean space, fin n → ℝ. -/
+instance standard_euclidean_space_is_vector_space (n : ℕ) : vector_space ℝ (fin n → ℝ) :=
+by apply_instance
+instance standard_euclidean_space_has_inner (n : ℕ) : has_inner (fin n → ℝ) :=
+{ inner := λ a b, ∑ i, a i * b i }
+instance standard_euclidean_space_is_inner_product_space (n : ℕ) :
+  inner_product_space (fin n → ℝ) :=
```

Comment:
Presumably these should move to wherever `inner_product_space` is defined, and probably be generalised from `fin n` to an arbitrary `fintype`.

Reply:
There is a problem here. The space `fin n → ℝ` has already a normed space instance, just like every finite product, given by the sup norm. Here, you are registering a second normed space instance on the very same space. This means that when you refer to the norm of an element of this space, Lean will have two choices and will sometimes choose one, sometimes the other, and this will become a mess. A good rule about typeclasses is that there should be at most one type class of a kind on a given type.

The way out here is to use a copy of `fin n → ℝ`, just like it is done in manifolds. Define
```
def euclidean_space (n : ℕ) : Type := fin n → ℝ
```
and then use this type. Here, something funny is going on: to Lean's kernel, this type is the same as `fin n → ℝ`, so if you have one of its elements `f` you can wite `f i` and compute as usual. But typeclass inference does not see through names, so it will not use the normed space instance of `fin n → ℝ` on `euclidean_space n`. Which means that you can put your own instance (here the inner product one) and this will not create ambiguities.

Of course, the vector space structure will also have to be declared, since typeclass inference will not see through the name. See how it is done in `geometry/manifold/real_instances.lean`.  

This space and its instances should probably be put in `inner_product_space`. Then you can also remove the definition in  `geometry/manifold/real_instances.lean`, and import `inner_product_space` instead (or let me do it in a future PR if you prefer).

Reply:
Does `geometry/manifold/real_instances` rely on the current choice of norm on `euclidean_space`?

Reply:
It probably does through the fact that continuity on a product is equivalent to continuity coordinate-wise. So, @jsm28 , don't touch it, I'll refactor later.

Reply:
Now defining `euclidean_space`, with instances as for manifolds, in `analysis/normed_space/real_inner_product.lean`.

Reply:
Note that this `euclidean_space` is still for `fin n`; I expect some other name should be used with a general `fintype`. Also, the linter fails when there are both this `euclidean_space` and the one for manifolds, so dealing with the manifold one later means we need to rename one of the two now (to what?).

Reply:
You can rename the manifold one to `euclidean_space2 n` for now. If we need the euclidean space indexed by an arbitrary fintype one day, then we will come up with a name then, but for now it doesn't seem to be needed.

Reply:
I've done that renaming. I didn't change the name of the lemma `findim_euclidean_space`; probably after the refactoring that lemma should end up in `analysis/normed_space/real_inner_product.lean`. And I didn't change the comments referring to `euclidean_space` in `smooth_manifold_with_corners.lean`.